src/TorchSharp/Torch.cs

```diff
@@ -304,26 +304,82 @@ public static partial class nn
             public static partial class utils
             {
                 [DllImport("LibTorchSharp")]
-                extern static double THSTensor_clip_grad_norm_(IntPtr tensor, int len, double max_norm, double norm_type);
+                extern static double THSTensor_clip_grad_norm_(IntPtr tensors, int len, double max_norm, double norm_type);
+
+                [DllImport("LibTorchSharp")]
+                extern static void THSTensor_clip_grad_value_(IntPtr tensors, int len, double clip_value);
+
+                [DllImport("LibTorchSharp")]
+                extern static IntPtr THSTensor_parameters_to_vector(IntPtr tensors, int len);
+
+                [DllImport("LibTorchSharp")]
+                extern static void THSTensor_vector_to_parameters(IntPtr vec, IntPtr tensors, int len);
 
                 /// <summary>
                 /// Clips gradient norm of an iterable of parameters.
                 /// The norm is computed over all gradients together, as if they were concatenated into a single vector.
-                /// Gradients are modified in-place.
                 /// </summary>
                 /// <param name="tensors"></param>
                 /// <param name="max_norm"></param>
                 /// <param name="norm_type"></param>
-                /// <returns></returns>
+                /// <remarks>Gradients are modified in-place.</remarks>
                 public static double clip_grad_norm_(IEnumerable<Modules.Parameter> tensors, double max_norm, double norm_type = 2.0)
+                {
+                    using (var parray = new PinnedArray<IntPtr>()) {
+                        IntPtr tensorsRef = parray.CreateArray(tensors.Select(p => p.Handle).ToArray());
+                        var value = THSTensor_clip_grad_norm_(tensorsRef, parray.Array.Length, max_norm, norm_type);
+                        CheckForErrors();
```

Comment:
>CheckForErrors [](http://example.com/codeflow?start=24&length=14)

Why we don't include the results of the CheckOfErrors inside the pinvoked method like THSTensor_clip_grad_norm_? 

I am asking as this can save pinvoke call. I am not expecting you change that in this PR but may be something we can track to improve?

Reply:
I'm not sure I understand what you are suggesting. The CheckForErrors() method transfers the native exception information, if there is one, to managed code and throws a managed-code exception.

Reply:
`CheckForErrors` is calling the following pinvoke method

```C#
            var error = THSTorch_get_and_reset_last_err();
```

You can have the native implementation of methods like `THSTensor_clip_grad_norm_` to call `THSTorch_get_and_reset_last_err()` and return the error as out parameter. Doing that, the code will look something like:

```C#
 var value = THSTensor_clip_grad_norm_(tensorsRef, parray.Array.Length, max_norm, norm_type out error);
 Throw_If_Error(error); // completely managed method
```

Reply:
Got it. That is a huge change, IMO. Maybe it's worth it, but definitely something we would separately, as you mention.